src/ray/object_manager/object_manager.cc

```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
 }
 
 ray::Status ObjectManager::Cancel(const ObjectID &object_id) {
-  ray::Status status = object_directory_->UnsubscribeObjectLocations(object_id);
+  ray::Status status = object_directory_->UnsubscribeObjectLocations(
+      object_directory_pull_callback_id_, object_id);
   return status;
 }
 
-ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
-                                uint64_t timeout_ms, int num_ready_objects,
+ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t wait_ms,
+                                uint64_t num_required_objects, bool wait_local,
                                 const WaitCallback &callback) {
-  // TODO: Implement wait.
+  UniqueID wait_id = UniqueID::from_random();
+
+  if (wait_local) {
+    return ray::Status::NotImplemented("Wait for local objects is not yet implemented.");
+  }
+
+  RAY_CHECK(wait_ms >= 0);
+  RAY_CHECK(num_required_objects != 0);
+  RAY_CHECK(num_required_objects <= object_ids.size());
+  if (object_ids.size() == 0) {
+    callback(std::vector<ObjectID>(), std::vector<ObjectID>());
+  }
+
+  // Initialize fields.
+  active_wait_requests_.emplace(wait_id, WaitState(*main_service_, wait_ms, callback));
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  wait_state.object_id_order = object_ids;
+  wait_state.wait_ms = wait_ms;
+  wait_state.num_required_objects = num_required_objects;
+  for (auto &oid : object_ids) {
+    if (local_objects_.count(oid) > 0) {
+      wait_state.found.insert(oid);
+    } else {
+      wait_state.remaining.insert(oid);
+    }
+  }
+
+  if (wait_state.remaining.empty()) {
```

Comment:
if we're going to include this special case here, then the condition should probably be something like `wait_state.found.size() >= num_required_objects || wait_ms == 0`.

That said, I think it makes sense to handle this case after firing off the `LookupLocations` calls.

Reply:
We invoke the lookup calls immediately after checking which objects are local to obtain current information about the location of remote objects. This allows us to collect information about all given objects, regardless of their location.

If we invoke when `wait_state.found.size() >= num_required_objects || wait_ms == 0` we may invoke `WaitComplete` before checking which remote objects are ready; we'll bias locally available objects.

If we fire `LookupLocations` and check `wait_state.found.size() >= num_required_objects || wait_ms == 0` independently, and the condition is `true`, the calls to `LookupLocations` will do nothing because the callback to `Wait` will already be invoked by the time any of the handlers for `LookupLocations` are invoked for the given `wait_id`.